crypto/fipsmodule/sha/keccak1600.c

```diff
@@ -394,21 +400,75 @@ void Keccak1600_Squeeze(uint64_t A[KECCAK1600_ROWS][KECCAK1600_ROWS], uint8_t *o
     }
 }
 
-#else
+#if defined(KECCAK1600_ASM)
+
+// Double-check that bit-interleaving is not used on AArch64
+#if BIT_INTERLEAVE != 0
+#error Bit-interleaving of Keccak1600 states should be disabled for AArch64
+#endif
 
-size_t Keccak1600_Absorb_hw(uint64_t A[KECCAK1600_ROWS][KECCAK1600_ROWS], const uint8_t *inp, size_t len,
-                       size_t r);
+#if defined(KECCAK1600_S2N_BIGNUM_ASM)
+#include "../../../third_party/s2n-bignum/s2n-bignum_aws-lc.h"
+
+// Dispatch logic for Keccak-x1 on AArch64:
+//
+// 1. If ASM is disabled, we use the C implementation.
+// 2. If ASM is enabled:
+//   - For Neoverse N1, V1, V2, we use scalar Keccak assembly from s2n-bignum,
+//     leveraging lazy rotations from https://eprint.iacr.org/2022/1243.
+//   - For Arm-based Apple CPUs, we use Neon Keccak assembly from s2n-bignum,
+//     leveraging the AArch64 SHA3 extension.
+//   - Otherwise, fall back to scalar Keccak implementation from OpenSSL,
+//     not using lazy rotations.
+//
+// Lazy rotations improve performance by up to 10% on CPUs with free
+// Barrel shifting, which includes Neoverse N1, V1, and V2. Not all
+// CPUs have free Barrel shifting (e.g. Apple M1 or Cortex-A72), so we
+// don't use it by default.
+//
+// Neoverse V1 and V2 do support SHA3 instructions, but they are only
+// implemented on 1/4 of Neon units, and are thus slower than a scalar
+// implementation.
+
+static int keccak_use_s2n_bignum_main(void) {
+    if (CRYPTO_is_Neoverse_N1() || CRYPTO_is_Neoverse_V1() || CRYPTO_is_Neoverse_V2()) {
+        return 1;
+    }
 
-size_t Keccak1600_Absorb(uint64_t A[KECCAK1600_ROWS][KECCAK1600_ROWS], const uint8_t *inp, size_t len,
-                   size_t r) {
-    return Keccak1600_Absorb_hw(A, inp, len, r);
+    return 0;
+}
+
+#if defined(MY_ASSEMBLER_SUPPORTS_NEON_SHA3_EXTENSION)
+static int keccak_use_s2n_bignum_alt(void) {
+    return (CRYPTO_is_ARMv8_SHA3_capable() && CRYPTO_is_ARMv8_Apple_M());
+}
+#endif // MY_ASSEMBLER_SUPPORTS_NEON_SHA3_EXTENSION
+#endif // KECCAK1600_S2N_BIGNUM_ASM
+
+// Scalar implementation from OpenSSL provided by keccak1600-armv8.pl
+extern void KeccakF1600_hw(uint64_t state[25]);
+
+void KeccakF1600(uint64_t A[KECCAK1600_ROWS][KECCAK1600_ROWS]) {
+#if defined(KECCAK1600_S2N_BIGNUM_ASM)
+#if defined(MY_ASSEMBLER_SUPPORTS_NEON_SHA3_EXTENSION)
```

Comment:
Are there other ways to arrange this to remove 1 or 2 levels of if/defs. At this point we are 3 levels deep into if/defs.
For other s2n-bignum integrations, we would typically stub out any missing functions.
I understand it's hard to always pass the s2n-bignum files intot he build if `sha3` is used in the asm implementation and not it's actual encoding.

Reply:
I'm not sure how to get rid of the #defines altogether, but am happy about concrete suggestions. For now, I think the following is already a bit better:
```c
void KeccakF1600(uint64_t A[KECCAK1600_ROWS][KECCAK1600_ROWS]) {
#if defined(KECCAK1600_S2N_BIGNUM_ASM) && defined(MY_ASSEMBLER_SUPPORTS_NEON_SHA3_EXTENSION)
    if (keccak_use_s2n_bignum_alt()) {
        sha3_keccak_f1600_alt((uint64_t *)A, iotas);
        return;
    }
#endif // KECCAK1600_S2N_BIGNUM_ASM && MY_ASSEMBLER_SUPPORTS_NEON_SHA3_EXTENSION

#if defined(KECCAK1600_S2N_BIGNUM_ASM)
    if (keccak_use_s2n_bignum_main()) {
        sha3_keccak_f1600((uint64_t *)A, iotas);
        return;
    }
#endif // KECCAK1600_S2N_BIGNUM_ASM

    KeccakF1600_hw((uint64_t *) A);
}
```